sway-lib-std/src/u256.sw

```diff
@@ -570,3 +606,63 @@ impl core::ops::Divide for U256 {
         quotient
     }
 }
+
+impl Power for U256 {
+    /// Fast exponentiation by squaring
+	/// https://en.wikipedia.org/wiki/Exponentiation_by_squaring
+	///
+	/// # Panics
+	///
+	/// Panics if the result overflows the type.
+    fn pow(self, exponent: Self) -> Self {
+        let mut value = self;
+        let mut exp = exponent;
+        let one = Self::one();
+
+        if exp.is_zero() {
+            return one;
+        }
+
+        if exp == one {
+            // Manually clone `self`. Otherwise, we may have a `MemoryOverflow`
+            // issue with code that looks like: `x = x.pow(other)`
+            return Self::from((self.a, self.b, self.c, self.d));
+        }
+
+        while (exp & one).is_zero() {
+            value = value * value;
+            exp >>= 1;
+        }
+
+        if exp == one {
+            return value;
+        }
+
+        let mut acc = value;
+        while exp > one {
+            exp >>= 1;
+            value = value * value;
+            if exp & one == one {
+                acc = acc * value;
+            }
+        }
+        acc
+    }
+}
+
+#[test]
+fn test_pow_u256() {
+    let five = U256::from((0, 0, 0, 5));
+    let two = U256::from((0, 0, 0, 2));
+    let three = U256::from((0, 0, 0, 3));
+    let twenty_eight = U256::from((0, 0, 0, 28));
+    assert_eq(five.pow(two), U256::from((0, 0, 0, 25)));
+    assert_eq(five.pow(three), U256::from((0, 0, 0, 125)));
+    assert_eq(five.pow(twenty_eight), U256::from((0, 359414837200037395, 18446744073709551615, 18446744073709551615)));
```

Comment:
I got from Wolfram: https://www.wolframalpha.com/input?i=5%5E28+%3D%3D+2*18446744073709551615%2B359414837200037395.

Another option is https://www.wolframalpha.com/input?i=convert+5%5E28+to+base+18446744073709551615%5D

```
input: convert 5^28 to base 18446744073709551615
Input interpretation
convert 37252902984619140625 to base18446744073709551615
Result
2:359414837200037395_18446744073709551615 (2 digits)
``

Reply:
Another good indication is: https://www.wolframalpha.com/input?i=log_5%5B18446744073709551615%5D

```
Input
log(5, 18446744073709551615)
Result
log(18446744073709551615)/log(5)
Decimal approximation
27.563299716697155242853137766599266959746869658989720909693829506...
```

Reply:
Using https://github.com/rust-num/num-bigint

![image](https://github.com/FuelLabs/sway/assets/83425/b93dac29-5121-4564-8fa0-5a0401283c01)

Reply:
> So `u64::MAX` is super close to `5^27`.

`u64::MAX` is 10996163476785723490 greater than 5^27 🤔

Reply:
> Ah, I see now. You're using the maximum value of `u64` as the base. Why is that?

Isn´t this what U256 is doing with four u64? Each field is a u64 digit.

Reply:
I suppose I misunderstood since their test cases seem to imply that every next `u64` is catching the overflow of the previous. I'm mostly asking out of curiosity since this area isn't my forte @xunilrj , sorry to have started such a long thread on your PR @andrewvious 😅 

This could imply a few different problems:
- `assert_eq` may be passing incorrectly (I think this is most likely, which could make the last problem true)
- the implementation of `Power` is somehow inaccurate
- the test case itself is inaccurate as @xunilrj stated

Reply:
No worries @eureka-cpu, I appreciate both of you @xunilrj I see how it works now, where on overflow it just increments the next u64 over by one. For example, adding 1 to `u64::MAX` would read:
``` 
u256 {
     u64=a: 0,
     u64=b: 0,
     u64=c: 1,
     u64=d: 0
} 
```